File: desktop/src/Beam.Core/Net/BlobChunkCodec.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using Beam.Core.Security;

namespace Beam.Core.Net;

/// <summary>
/// Codifica/decodifica frames binários <c>blob.chunk</c> com cifra AES-GCM por chunk.
/// Layout: blobId(16) | seq(uint32 BE) | len(uint32 BE) | pacote GCM.
/// </summary>
public static class BlobChunkCodec
{
    public const int HeaderSize = 24;

    public static byte[] EncodeChunk(ReadOnlySpan<byte> blobId, uint sequence, ReadOnlySpan<byte> plaintext, SessionCipher cipher)
    {
        if (blobId.Length != 16)
        {
            throw new ArgumentException("O blobId deve ter 16 bytes.", nameof(blobId));
        }

        var encrypted = cipher.Encrypt(plaintext, AssociatedData(blobId, sequence));
        var frame = new byte[HeaderSize + encrypted.Length];
        blobId.CopyTo(frame);
        BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(16, 4), sequence);
        BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(20, 4), (uint)encrypted.Length);
        encrypted.CopyTo(frame.AsSpan(HeaderSize));
        return frame;
    }

    public static bool TryDecodeChunk(
        ReadOnlySpan<byte> frame,
        SessionCipher cipher,
        out byte[] blobId,
        out uint sequence,
        out byte[] plaintext)
    {
        blobId = [];
        sequence = 0;
        plaintext = [];

        if (frame.Length < HeaderSize)
        {
            return false;
        }

        blobId = frame[..16].ToArray();
        sequence = BinaryPrimitives.ReadUInt32BigEndian(frame.Slice(16, 4));
        var payloadLength = BinaryPrimitives.ReadUInt32BigEndian(frame.Slice(20, 4));
        if (payloadLength == 0 || frame.Length < HeaderSize + payloadLength)
        {
            return false;
        }

        try
        {
            plaintext = cipher.Decrypt(frame.Slice(HeaderSize, (int)payloadLength), AssociatedData(blobId, sequence));
            return true;
        }
        catch (CryptographicException)
        {
            return false;
        }
    }

    public static byte[] CreateBlobIdBytes() => Guid.NewGuid().ToByteArray();

    public static string ToBlobIdString(ReadOnlySpan<byte> blobIdBytes) =>
        Convert.ToHexStringLower(blobIdBytes);

    public static byte[] ParseBlobIdString(string blobId)
    {
        if (blobId.Length != 32)
        {
            throw new ArgumentException("O blobId deve ter 32 caracteres hex.", nameof(blobId));
        }

        return Convert.FromHexString(blobId);
    }

    public static string ComputeSha256Hex(ReadOnlySpan<byte> data) =>
        Convert.ToHexStringLower(SHA256.HashData(data));

    private static byte[] AssociatedData(ReadOnlySpan<byte> blobId, uint sequence)
    {
        var aad = new byte[blobId.Length + 4];
        blobId.CopyTo(aad);
        BinaryPrimitives.WriteUInt32BigEndian(aad.AsSpan(blobId.Length, 4), sequence);
        return aad;
    }
}

File: mobile/app/src/main/java/com/esousa/beam/blob/BlobChunkCodec.kt
package com.esousa.beam.blob

import com.esousa.beam.security.SessionCipher
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.UUID

object BlobChunkCodec {
    const val HEADER_SIZE = 24

    fun encodeChunk(blobId: ByteArray, sequence: Int, plaintext: ByteArray, cipher: SessionCipher): ByteArray {
        require(blobId.size == 16) { "O blobId deve ter 16 bytes." }
        val encrypted = cipher.encrypt(plaintext, associatedData(blobId, sequence))
        return ByteBuffer.allocate(HEADER_SIZE + encrypted.size).apply {
            put(blobId)
            putInt(sequence)
            putInt(encrypted.size)
            put(encrypted)
        }.array()
    }

    fun decodeChunk(frame: ByteArray, cipher: SessionCipher): DecodedChunk? {
        if (frame.size < HEADER_SIZE) return null
        val buffer = ByteBuffer.wrap(frame)
        val blobId = ByteArray(16).also { buffer.get(it) }
        val sequence = buffer.int
        val payloadLength = buffer.int
        if (payloadLength <= 0 || frame.size < HEADER_SIZE + payloadLength) return null
        val encrypted = frame.copyOfRange(HEADER_SIZE, HEADER_SIZE + payloadLength)
        return runCatching {
            DecodedChunk(
                blobId = blobId.toHex(),
                sequence = sequence,
                plaintext = cipher.decrypt(encrypted, associatedData(blobId, sequence)),
            )
        }.getOrNull()
    }

    fun createBlobIdBytes(): ByteArray = uuidToBytes(UUID.randomUUID())

    fun sha256Hex(data: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(data).toHex()

    private fun associatedData(blobId: ByteArray, sequence: Int): ByteArray =
        ByteBuffer.allocate(blobId.size + 4).apply {
            put(blobId)
            putInt(sequence)
        }.array()

    private fun uuidToBytes(uuid: UUID): ByteArray =
        ByteBuffer.allocate(16).apply {
            putLong(uuid.mostSignificantBits)
            putLong(uuid.leastSignificantBits)
        }.array()

    private fun ByteArray.toHex(): String = joinToString(separator = "") { byte ->
        "%02x".format(byte)
    }
}

data class DecodedChunk(val blobId: String, val sequence: Int, val plaintext: ByteArray)
